### PR TITLE
feat(hikes): error handling `HikesViewModel`

### PR DESCRIPTION
**Pull request by MaelImhof**
# Summary

This PR introduces better error handling features in `HikesViewModel`. Mainly, a new `DeferredData.Error` type has been created to represent a request that happened but failed, and the new state flow `loadingErrorMessageId` was added in the view model to tell the UI when an error occurred, and which message to display.

# Details

There aren't that many places where errors need to be handled:

- **Map screen**
   - If an error occurs when searching for hikes in bounds (`Search hikes here`), a message is displayed instead of the hikes list
   - If an error occurs while loading the elevation of a hike, an empty list will be passed to the `HikeCard` to display "No data" on the graph
- Saved hikes screen
   - If an error occurs when loading the saved hikes or their OSM data, an error message is displayed with a retry button
   - If an error occurs while loading the elevation of a hike, an empty list will be passed to the `HikeCard` to display "No data" on the graph
- Details screen + Run hike screen
   - If an error occurs while loading the selected hike's complete information, an error screen will be displayed with a retry button

### CODE REVIEW

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt [hunk 1]
```diff
@@ -166,21 +168,41 @@ class RunHikeScreenTest {
...
         .assertIsDisplayed()
+        .assertHasClickAction()
+        .performClick()
+
+    // Clicking the button should trigger going back with the navigation
+    verify(mockNavigationActions, times(1)).goBack()
   }
 
   @Test
   fun runHikeScreen_displaysLoadingAnimationForElevation() = runTest {
```
Comment (davidholzwarth) [inline app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:190]: Nitpick: This test's name is not very descriptive, it think it would be more understandable if it were called `runHikeScreen_displaysLoadingAnimationForElevation_whenElevatrionRetrievalFails`

File: app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt [hunk 2]
```diff
@@ -101,6 +105,68 @@ class SavedHikesScreenTest : TestCase() {
...
+        .onNodeWithTag(CenteredErrorAction.TEST_TAG_CENTERED_ERROR_MESSAGE)
+        .assertIsDisplayed()
+    composeTestRule
+        .onNodeWithTag(CenteredErrorAction.TEST_TAG_CENTERED_ERROR_BUTTON)
+        .assertIsDisplayed()
+        .assertHasClickAction()
+        .performClick()
+
+    // Check that the repository was called again, because the button is a retry action
+    verify(savedHikesRepository, times(2)).loadSavedHikes()
```
Comment (davidholzwarth) [inline app/src/androidTest/java/ch/hikemate/app/ui/saved/SavedHikesScreenTest.kt:167]: Praise: Nice test!

File: app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt [hunk 3]
```diff
@@ -421,8 +422,15 @@ class HikesViewModel(
...
+   *
+   * Additionally, updates [_loadingErrorMessageId] with the [errorMessageId] parameter.
+   *
+   * @param value The new value of the loading state, whether loading (true) or not (false).
+   * @param errorMessageId The ID of the string resource to display to the user if the loading
+   *   operation has failed. This parameter is only considered if [value] is false. This value will
+   *   be set even if other loading operations are still ongoing.
    */
-  private suspend fun setLoading(value: Boolean) =
+  private suspend fun setLoading(value: Boolean, errorMessageId: Int?) =
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt:433]: I had this debate while writing the function. The thing is, if you set `null` as default value and someone forgets to put an error message for some reason, then it will interpret it as a success.

Furthermore, this function is internal to the view model, it's not used outside of it, so I assumed the users of this function will "know" it and understand they have to pass an error message id to make it an error.

In the end, I personally think it might be better to require the user of the function to explicitly say "there was no problem, take `null` as parameter".

Given my thinking process on this one, what's your call? I can change it if you still believe I should.

File: app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt [hunk 4]
```diff
@@ -198,15 +213,23 @@ private fun SavedHikes(savedHikes: List<StateFlow<Hike>>?, hikesViewModel: Hikes
...
-  // Ask for elevation if not already available
-  if (!hike.elevation.obtained()) {
+  val elevation: List<Double>?
+  if (hike.elevation is DeferredData.Error) {
+    // Display an empty elevation graph if the data is not available because of an error
+    elevation = emptyList()
+  } else if (!hike.elevation.obtained()) {
+    // Ask for elevation if not already available
     hikesViewModel.retrieveElevationDataFor(hike.id)
+    elevation = null
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/ui/saved/SavedHikesScreen.kt:223]: Yes, the composable will be recomposed every time the hike changes. In particular, here the elevation is not yet available, and `null` is the value to pass to the graph composable for it to display `Loading data...`, so I pass this after requesting the elevation data to the view model.

And once the elevation data comes back, this will be recomposed.

File: app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt [hunk 5]
```diff
@@ -385,29 +400,31 @@ class HikesViewModelTest {
...
       runTest(dispatcher) {
-        // Listen to the changes made to loading during the call
-        val emissions = mutableListOf<Boolean>()
-        val job = backgroundScope.launch { hikesViewModel.loading.collect { emissions.add(it) } }
+        // Before launching the operation, loading should be false
+        assertFalse(hikesViewModel.loading.value)
 
         // Set the repository to throw an exception because we do not care
-        coEvery { savedHikesRepo.loadSavedHikes() } throws Exception("Failed to load saved hikes")
+        coEvery { savedHikesRepo.loadSavedHikes() } answers
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:414]: Thank you for your warning, although I'm not sure how exactly this case could happen. Were you able to observe it?

`runTest` is used with a `dispatcher` parameter. The dispatcher is an `UnconfinedTestDispatcher`, meaning it runs all the coroutines started on it pretty much instantaneously. The test shouldn't end before all of its coroutines are done. The same dispatcher is provided to the `HikesViewModel` instance and used as main dispatcher, using

```kotlin
Dispatchers.setMain(dispatcher)
```

So basically, as far as I understand, everyone uses the same scheduler and all coroutines should terminate before the end of the test.

Perhaps you could point me to the fix the site proposes to solve this? I'm not sure what I should look for on the link you provided.

File: app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt [hunk 6]
```diff
@@ -507,29 +532,31 @@ class HikesViewModelTest {
...
+            {
+              // Check that during the operation, loading is set to true
+              assertTrue(hikesViewModel.loading.value)
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:549]: See my answer in the first comment of this type.

File: app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt [hunk 7]
```diff
@@ -1178,25 +1215,25 @@ class HikesViewModelTest {
...
+        // Before launching the operation, loading should be false
+        assertFalse(hikesViewModel.loading.value)
 
         // Set the repository to throw an exception because we do not care
-        coEvery { osmHikesRepo.getRoutes(any(), any(), any()) } throws
-            Exception("Failed to load saved hikes")
+        coEvery { osmHikesRepo.getRoutes(any(), any(), any()) } answers
+            {
+              // Check that during the operation, loading is set to true
+              assertTrue(hikesViewModel.loading.value)
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:1225]: See my answer in the first comment of this type.

File: app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:414]: So a way you can test it is by:

1. Delete those lines in `HikesViewModel.refreshSavedHikesCache`:
```kotlin
// Let the user know a heavy load operation is being performed
setLoading(value = true, errorMessageId = null)
```
2. Run the test `refreshSavedHikesCache sets loading to true then false`

The test should fail, because loading is not set to true in the function.

[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:414]: But you are right, I should add something that checks whether the coroutine has terminated.

[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:414]: I just realized, I actually already do that, on this line :

```kotlin
// Check that the operation occurred, i.e. the repo was called
coVerify(exactly = 1) { savedHikesRepo.loadSavedHikes() }
```

This effectively checks that the function was called and terminated, doesn't it? Maybe it doesn't check that it terminated, but I'm not sure I need to add an additional check. What do you think?

File: app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt [hunk 8]
```diff
@@ -1514,6 +1561,39 @@ class HikesViewModelTest {
...
+
+        // The elevation repository should be called exactly once
+        coVerify(exactly = 1) { elevationRepo.getElevation(any(), any(), any()) }
+        // The appropriate callback should be called
+        assertTrue(onFailureCalled)
+        // The hike should now have its elevation data set to Error
+        assertEquals(1, hikesViewModel.hikeFlows.value.size)
+        assertEquals(singleOsmHike1[0].id, hikesViewModel.hikeFlows.value[0].value.id)
+        assertTrue(hikesViewModel.hikeFlows.value[0].value.elevation is DeferredData.Error)
+      }
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt:1595]: Thank you ! 😄

File: app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt [hunk 9]
```diff
@@ -178,6 +178,23 @@ class MapScreenTest : TestCase() {
...
+      val onFailure = it.getArgument<(Exception) -> Unit>(2)
+      onFailure(Exception("Test exception"))
+    }
+
+    // Perform a search
+    composeTestRule.onNodeWithTag(MapScreen.TEST_TAG_SEARCH_BUTTON).performClick()
+
+    // Check that the error message is displayed
+    composeTestRule.onNodeWithTag(MapScreen.TEST_TAG_LOADING_ERROR_MESSAGE).assertIsDisplayed()
+  }
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt:196]: Thank you 😄

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt [hunk 10]
```diff
@@ -166,21 +168,41 @@ class RunHikeScreenTest {
...
+    composeTestRule.onNodeWithTag(CenteredErrorAction.TEST_TAG_CENTERED_ERROR_MESSAGE)
     composeTestRule
-        .onNodeWithTag(CenteredLoadingAnimation.TEST_TAG_CENTERED_LOADING_ANIMATION)
+        .onNodeWithTag(CenteredErrorAction.TEST_TAG_CENTERED_ERROR_BUTTON)
         .assertIsDisplayed()
+        .assertHasClickAction()
+        .performClick()
+
+    // Clicking the button should trigger a retry of the elevation retrieval
+    verify(elevationRepository, times(2)).getElevation(any(), any(), any())
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:205]: Hehe I know right? 😎 😂

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt [hunk 11]
```diff
@@ -166,21 +168,41 @@ class RunHikeScreenTest {
   }
 
   @Test
-  fun runHikeScreen_displaysLoadingAnimationForWaypoints() = runTest {
+  fun runHikeScreen_displaysErrorMessageForWaypoints() = runTest {
```
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:171]: True, my name is not very descriptive, although your suggestion seems a bit long. I settled for

```
runHikeScreen_displaysError_whenWaypointsRetrievalFails
```

which looks like a good compromise to me. I also changed similar tests. Let me know what you think!

File: app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt (hunk 1 above)
Comment (MaelImhof) [reply app/src/androidTest/java/ch/hikemate/app/ui/map/RunHikeScreenTest.kt:190]: Same as above, settled for a name in the middle: more descriptive and shorter. Let me know what you think!

**Review by davidholzwarth — COMMENTED** [review]
Just some comments left to change

File: app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaelImhof) [reply app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt:434]: Yes, I realize the documentation is not very clear, I added some to further clarify what this parameter does.
